README.rst

```diff
@@ -1,334 +1,503 @@
-
-    @pytest.mark.datafiles(
-        os.path.join(FIXTURE_DIR, 'dir1'),
-        os.path.join(FIXTURE_DIR, 'dir2'),
-        os.path.join(FIXTURE_DIR, 'dir3'),
-        on_duplicate='replace',
-        )
-    def test_dir_replace(datafiles):
-        # datafiles.listdir() will list fileA and fileB originally from dir3
-        pass
-
-    @pytest.mark.datafiles(
-        os.path.join(FIXTURE_DIR, 'dir1'),
-        os.path.join(FIXTURE_DIR, 'dir2'),
-        os.path.join(FIXTURE_DIR, 'dir3'),
-        # on_duplicate='exception' is the default and does not need to be
-        # specified
-        )
-    def test_dir_exception(datafiles):
-        # An exception will be raised because of duplicate filename fileA
-        pass
-
-    @pytest.mark.datafiles(
-        os.path.join(FIXTURE_DIR, 'dir1'),
-        os.path.join(FIXTURE_DIR, 'dir2'),
-        os.path.join(FIXTURE_DIR, 'dir3'),
-        keep_top_dir=True,
-        )
-    def test_dir_keep_top_dir(datafiles):
-        # datafiles.listdir() will list dir1, dir2 and dir3 (each containing
-        # fileA and fileB)
-        pass
+---------
+
+Imagine you have 3 directories (*dir1*, *dir2*, *dir3*) each
+containing the files (*fileA* and *fileB*).
+
+This example clarifies the options **on_duplicate** and
+**keep_top_dir**.
+
+.. code:: python
+
+   """Example: files with same names."""
+   from pathlib import Path
+
+   import pytest
+
+   FIXTURE_DIR = Path(__file__).parent.resolve() / '_fixture_files'
+
+
+   @pytest.mark.datafiles(
+       FIXTURE_DIR / 'dir1',
+       FIXTURE_DIR / 'dir2',
+       FIXTURE_DIR / 'dir3',
+       on_duplicate='ignore',
+   )
+   def test_dir_ignore(datafiles):
+       """Use files from dir1 (first dir added)."""
+       assert len(list(datafiles.iterdir())) == 2
+       assert (datafiles / 'fileA').exists()
+       assert (datafiles / 'fileA').read_text() == '1a\n'
+
+
+   @pytest.mark.datafiles(
+       FIXTURE_DIR / 'dir2',
+       FIXTURE_DIR / 'dir1',
+       FIXTURE_DIR / 'dir3',
+       on_duplicate='ignore',
+   )
+   def test_dir_ignore2(datafiles):
+       """Use files from dir2 (first dir added)."""
+       assert len(list(datafiles.iterdir())) == 2
+       assert (datafiles / 'fileA').exists()
+       assert (datafiles / 'fileA').read_text() == '2a\n'
+
+
+   @pytest.mark.datafiles(
+       FIXTURE_DIR / 'dir1',
+       FIXTURE_DIR / 'dir2',
+       FIXTURE_DIR / 'dir3',
+       on_duplicate='overwrite',
+   )
+   def test_dir_overwrite(datafiles):
+       """Use files from dir3 (last dir added)."""
+       assert len(list(datafiles.iterdir())) == 2
+       assert (datafiles / 'fileA').exists()
+       assert (datafiles / 'fileA').read_text() == '3a\n'
+
+
+   @pytest.mark.datafiles(
+       FIXTURE_DIR / 'dir1',
+       FIXTURE_DIR / 'dir2',
+       FIXTURE_DIR / 'dir3',
+       # on_duplicate='exception' is the default
+   )
+   @pytest.mark.skip(
+       reason='will raise an exception that cannot be caught in the test itself'
+   )
```

Comment:
This is again a case where you have to make adjustments to make the examples runnable and are sacrificing readability of the examples.

How about the following solution: Write a (python) script that executes pytest on the examples and ensures some pre-requisites and verifies the expected result.

For example this script could replace the string /opt/big_files with /tmp/big_files and create that directory. The script could also verify that a test is raising an exception.

Some pseudo-code:

```python
#!/usr/env python3

# Execute example1.py

# Read in the file
with open('exampl1.py', 'r') as file :
    filedata = file.read()

# Replace the target string
filedata = filedata.replace('/opt/big_files', '/tmp/big_files')

# Write the file out again
with open('example1.py', 'w') as file:
    file.write(filedata)

process = subprocess.Popen(
    ['pytest', 'example1.py'],
    stdout=subprocess.PIPE, 
    stderr=subprocess.PIPE,
)
stdout, stderr = process.communicate()
if stderr:
    raise ValueError("stderr should have been empty. example1 no longer works")
```

Reply:
nice one.

Reply:
implemented using pytest